### DISCUSSION TITLE
Why does tokio::sync::broadcast::channel not have a Send requirement?

### DISCUSSION THREAD

**Discussion by pretzelhammer** [tokio-rs/tokio#6459] [Q&A] [answered]
The function signature of `tokio::sync::broadcast::channel` is this:

```rust
pub fn channel<T: Clone>(capacity: usize) -> (Sender<T>, Receiver<T>)
```

But why is it `T: Clone` and not `T: Clone + Send`? Say I create a broadcast channel with something like:

```rust
let (tx, rx) = broadcast::channel::<Rc<RefCell<String>>(8);
```
And then I spawn a bunch of tasks give each of them an `rx`, and then I `tx.send(Rc::new(RefCell::new(String::from("whatever")))`, and then each of the tasks calls `clone` and `borrow_mut` a bunch of times on the received value concurrently... wouldn't that cause some kind of data race or undefined behavior?

**Comment by sfackler** [comment, marked as answer]
Sender and Receiver are only Send if T is.

**Reply by pretzelhammer** [reply]
huh, it never occurred to me that a broadcast channel could be in a single-threaded context, but that makes sense, thanks for clarifying!
